perftest/wallet.go

```diff
@@ -0,0 +1,425 @@
+	}
+	return pubKeys, nil
+}
+
+func createOrLoadWallets(walletURL string, number int) (int, error) {
+	// We want to make or load a set of wallets, do it in a loop here
+	var key string
+	var newKeys = 0
+	for i := 0; i < number; i++ {
+		userName := fmt.Sprintf("User%04d", i)
+
+		// Attempt to log in with that username
+		token, err := loginWallet(walletURL, userName, "p3rfb0t")
+		if err != nil {
+			token, err = createWallet(walletURL, userName, "p3rfb0t")
+			if err != nil {
+				return 0, fmt.Errorf("unable to create a new wallet: %w", err)
+			}
+		}
+		keys, _ := listKeys(walletURL, token)
+		if len(keys) == 0 {
+			key, _ = createKey(walletURL, userName, "p3rfb0t", token)
+			newKeys++
+		} else {
+			key = keys[0]
+		}
+
+		users = append(users, UserDetails{
+			userName: userName,
+			token:    token,
+			pubKey:   key,
+		})
+	}
+	return newKeys, nil
+}
+
+func signSubmitTx(user int, command string) error {
+	err := sendCommand([]byte(command), users[user].token)
+	return err
+}
+
+func sendCommand(submission []byte, token string) error {
+	postBuffer := bytes.NewBuffer(submission)
+
+	URL := "http://" + savedWalletURL + "/api/v1/command"
+
+	client := &http.Client{}
+	req, err := http.NewRequest(http.MethodPost, URL, postBuffer)
+	if err != nil {
+		return err
+	}
+
+	bearer := "Bearer " + token
+	req.Header.Add("Authorization", bearer)
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+	sb := string(body)
+	if strings.Contains(sb, "error") {
+		log.Println(sb)
+		return fmt.Errorf(sb)
+	}
+	return nil
+}
+
+func sendOrder(marketID string, user int, price, size int64,
+	orderType string, tif proto.Order_TimeInForce, expiresAt int64) {
+	cmd := `{ "orderSubmission": {
+      "marketId": "$MARKETID",
+      $PRICE
+      "size": "$SIZE",
+      "side": "$SIDE",
+      "timeInForce": "$TIME_IN_FORCE",
+      $EXPIRES_AT
+      "type": "$TYPE",
+      "reference": "order_ref"
+    },
+    "pubKey": "$PUBKEY",
+    "propagate": true
+  }`
+
+	cmd = strings.Replace(cmd, "$MARKETID", marketID, 1)
+	cmd = strings.Replace(cmd, "$SIZE", fmt.Sprintf("%d", abs(size)), 1)
+
+	if orderType == "LIMIT" {
+		cmd = strings.Replace(cmd, "$PRICE", fmt.Sprintf("\"price\": \"%d\",", price), 1)
+		cmd = strings.Replace(cmd, "$TYPE", "TYPE_LIMIT", 1)
+	} else {
+		cmd = strings.Replace(cmd, "$PRICE", "", 1)
+		cmd = strings.Replace(cmd, "$TYPE", "TYPE_MARKET", 1)
+	}
+
+	if expiresAt > 0 {
+		cmd = strings.Replace(cmd, "$EXPIRES_AT", fmt.Sprintf("\"expiresAt\": %d,", expiresAt), 1)
+	} else {
+		cmd = strings.Replace(cmd, "$EXPIRES_AT", "", 1)
+	}
+
+	if size > 0 {
+		cmd = strings.Replace(cmd, "$SIDE", "SIDE_BUY", 1)
+	} else {
+		cmd = strings.Replace(cmd, "$SIDE", "SIDE_SELL", 1)
+	}
+
+	switch tif {
+	case proto.Order_TIME_IN_FORCE_GTT:
+		cmd = strings.Replace(cmd, "$TIME_IN_FORCE", "TIME_IN_FORCE_GTT", 1)
```

Comment:
if we use a struct as suggested above, the field values should just be set as e.g. `proto.Order_TIME_IN_FORCE_GTT` , so there would be no need for this switch